packages/@react-spectrum/s2/src/SegmentedControl.tsx

```diff
@@ -0,0 +1,289 @@
+  let {
+    defaultValue,
+    value
+  } = props;
+  let domRef = useDOMRef(ref);
+
+  let prevRef = useRef<DOMRect>(null);
+  let currentSelectedRef = useRef<HTMLDivElement>(null);
+
+  let onChange = (value: string) => {
+    if (currentSelectedRef.current) {
+      prevRef.current = currentSelectedRef?.current.getBoundingClientRect();
+    }
+    
+    if (props.onChange) {
+      props.onChange(value);
+    }
+  };
+
+  return (
+    <RadioGroup 
+      {...props}
+      ref={domRef}
+      orientation="horizontal"
+      style={props.UNSAFE_style}
+      onChange={onChange}
+      className={(props.UNSAFE_className || '') + segmentedControl({size: 'M'}, props.styles)}
+      aria-label={props['aria-label']}>
+      <DefaultSelectionTracker defaultValue={defaultValue} value={value} prevRef={prevRef} currentSelectedRef={currentSelectedRef}>
+        {props.children}
+      </DefaultSelectionTracker>
+    </RadioGroup>
+  );
+}
+
+function DefaultSelectionTracker(props: DefaultSelectionTrackProps) {
+  let state = useContext(RadioGroupStateContext);
+  let isRegistered = useRef(!(props.defaultValue == null && props.value == null));
+  let disabledIsRegistered = useRef<string | null>(props.defaultValue ?? props.value ?? null);
+
+  // default select the first available item
+  let register = useCallback((value: string, isDisabled?: boolean) => {
+    if (state && !isRegistered.current && !isDisabled) {
+      isRegistered.current = true;
+
+      state.setSelectedValue(value);
+    } else if (isDisabled && disabledIsRegistered.current == null) {
+      disabledIsRegistered.current = value;
+    }
+  }, []);
+
+  // if the registration fails, then we will default select the first item
+  useLayoutEffect(() => {
+    if (state && isRegistered.current === false) {
+      state.setSelectedValue(disabledIsRegistered.current);
+    } else if (isRegistered.current === false && disabledIsRegistered.current == null) {
+      throw new Error('Could not determine a default selected item');
+    }
+  }, []);
+
+
+  return (
+    <Provider
+      values={[
+        [InternalSegmentedControlContext, {register: register, prevRef: props.prevRef, currentSelectedRef: props.currentSelectedRef}]
+      ]}> 
+      {props.children}
+    </Provider>
+  );
+}
+
+function SegmentedControlItem(props: SegmentedControlItemProps, ref: FocusableRef<HTMLLabelElement>) {
+  let inputRef = useRef<HTMLInputElement>(null);
+  let domRef = useFocusableRef(ref, inputRef);
+  let divRef = useRef<HTMLDivElement>(null);
+  let {register, prevRef, currentSelectedRef} = useContext(InternalSegmentedControlContext);
+  let state = useContext(RadioGroupStateContext);
+  let isSelected = props.value === state?.selectedValue;
+  // do not apply animation if a user has the prefers-reduced-motion setting
+  let isReduced = false;
+  if (window?.matchMedia) {
+    isReduced = window.matchMedia('(prefers-reduced-motion: reduce)').matches;
+  }
+  
+  let isIconOnly = useRef<boolean>(false);
```

Comment:
I think you could do this in CSS only with no JS if you use `:has()`. I believe we do that in button.

Reply:
yeah i think button does something like this `:has([slot=icon]:only-child)': 'control'`

but i don't think it would work in this scenario because there's a hidden input so it looks like
```
<div>
  <span with the visually hidden input>
  <icon>
</div>

```
so it's not really an only-child

Reply:
You could try something like this: `':has([slot=icon]:nth-child(2):nth-last-child(1))'`